### PR TITLE
[Bugfix][Spec Decode][Structured Output] DSpark: fix the grammar bitmask mapping when the draft budget is zero

### PR DESCRIPTION
**Pull request by oops-oom**
## Purpose

Fixes a crash when DSpark adaptive verification (`enable_adaptive_verification`, added in #47808) is combined with structured outputs and the chosen draft budget is zero.

```
assert num_masks == len(mapping)  // fails when draft_budget == 0
```

**Scenario:** serving a structured-outputs request (JSON schema / grammar) with DSpark adaptive verification enabled crashes on an assert as soon as the drafter's confidence drops far enough that adaptive verification decides to verify zero drafts this step.

### When is `draft_budget == 0`? (not an edge case)

That is not an exotic corner. Zero is the routine choice whenever drafts stop paying for their verification cost. Per step, the controller picks `draft_budget = argmax_b (est_accepted_tokens(b)/cost(b))` over `b ∈ [0, max_draft_budget]` (`adaptive_verification.py:329`). `b = 0` ("verify no drafts, sample only the bonus token") is a normal output of that argmax, selected whenever verifying even the most-confident draft is net-negative for throughput — i.e. whenever the drafter's confidence is low. That happens routinely (high-entropy tokens, warm-up, grammar-constrained steps), so a sustained DSpark + structured-outputs workload hits it repeatedly and the crash reproduces under ordinary load. This is exactly the behavior #47808 introduced.

### Root cause

The scheduler and the worker disagree on how many grammar bitmask rows exist:

  - The scheduler sizes the bitmask from the **scheduled** drafts — `len(drafts) + 1` rows per request (`vllm/v1/structured_output/__init__.py`), and it has no knowledge of the device-side budget trimming.
  - `compact_batch()`'s zero-budget branch rewrites `cu_num_logits_np` to the exact bonus-only layout (`adaptive_verification.py:361`). That rewrite exists to serve `_iter_request_chunks`, which slices the compacted logits with these offsets.

 `apply_grammar_bitmask` derived its bitmask → logits mapping from those rewritten offsets, so it produced one row per request instead of one per scheduled draft and tripped `assert num_masks == len(mapping)` (`structured_outputs.py:102`).

For a batch of 2 verification requests with 2 drafts each plus 1 bonus token, the scheduler emits 6 rows while the old expression yields 2:

 ```
 bitmask rows = 6,  mapping = [0, 3]  (len 2)  ->  assert fails
 ```

 ### Why the other two budget regimes were unaffected

Only the zero-budget branch rewrites `cu_num_logits_np`; the other two return it untouched, still holding the scheduled layout that the scheduler used. Exercising all three branches through the real `compact_batch`:

 | regime | `cu_num_logits_np` | rewritten | old mapping len | bitmask rows | |
 |---|---|---|---|---|---|
 | `budget == num_drafts` | `[0, 3, 6]` | no | 6 | 6 | OK |
 | `0 < budget < num_drafts` | `[0, 3, 6]` | no | 6 | 6 | OK |
 | `budget == 0` | `[0, 1, 2]` | **yes** | **2** | 6 | **assert fails** |

 So the old code was correct in two regimes only because that field happened to carry scheduled-layout semantics there.

 ### Fix

Size the mapping from `num_draft_tokens_per_req + num_bonus_tokens` instead. That array comes from the same `scheduled_spec_decode_tokens` the scheduler used (`model_runner.py:1132`) and is never touched by compaction, so all three regimes agree and the row count no longer depends on a field whose meaning shifts.

 Rows that the compacted device layout has no room for are already masked by  `position_is_active` in the kernel (`structured_outputs.py:144`), which resolves true per-request offsets from the GPU-side `cu_num_logits`. This is the mechanism #47808 introduced; the fix just stops the host side from second-guessing it.

 ## Test Plan

 Added `test_zero_budget_keeps_one_grammar_row_per_scheduled_draft` to `tests/v1/spec_decode/test_adaptive_verification.py`. 

  ```bash
  .venv/bin/python -m pytest tests/v1/spec_decode/test_adaptive_verification.py -v
  # related suites touching the same cu_num_logits_np contract:
  .venv/bin/python -m pytest tests/v1/worker/test_gpu_rejection_sampler_chunking.py \
                             tests/v1/worker/test_gpu_batch_ordering.py -v
  pre-commit run --all-files
  ```

  ## Test Result

  ```
  tests/v1/spec_decode/test_adaptive_verification.py  6 passed
  + test_gpu_rejection_sampler_chunking.py, test_gpu_batch_ordering.py
                                                    17 passed, 4 skipped
  pre-commit                                        all hooks passed (incl. mypy, ruff format)
  ```

  Verified the new test actually fails without the fix by reverting only the helper body while keeping its signature:

  ```
  without fix:  E   assert 3 == 7
                    where 3 = len([0, 3, 6])
  with fix:     6 passed

### CODE REVIEW

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for the PR; great catch! im wondering if theres a way to fix this without having to branch on `num_draft_tokens_per_req` experimenting with that here: https://github.com/vllm-project/vllm/pull/52477 but overall I think this approach is reasonable (we can land this in the interim)

**Review by LucasWilkinson — APPROVED** [review]

**Comment by oops-oom** [comment]
@LucasWilkinson Thanks for the review, and for the `ready` label. Let me read through #52477 before answering the `num_draft_tokens_per_req` question properly.
